### PR TITLE
WT-6136 Record incremental extent lists before merging them with earlier

### PR DESCRIPTION
**Pull request by sueloverso**
checkpoints.
Add reproducer test that reads all blocks returned to make sure they
have changes.

### CODE REVIEW

**Comment by sueloverso** [comment]
@keitharnoldsmith and @keithbostic this failed an incremental backup test so it is not ready yet. I've moved the branch into draft mode.

**Comment by agorrod** [comment]
@sueloverso and @keitharnoldsmith the testing failure here looks a little like WT-6118 to me - I wonder if this change means we see that issue more often (and possibly even without WT-5242)?

**Comment by keitharnoldsmith** [comment]
Hmm.  Interesting thought.  The initial symptom does look similar --- a
checksum error while reading a block from backup.

On one hand, we are both looking at scenarios where we need to make sure
that a correct and consistent set of data is captured and backed up.
Pretty much any bug where one or more blocks get left out would manifest as
a checksum error.

On the other hand, I think we're both relying on the existing checkpoint
code to accurately determine the set of blocks that were dirtied as part of
the checkpoint.  So if there's a bug in that code, it could be causing both
of our problems.

(All this assuming that my quick reading of WT-6136 was accurate)

Good job arranging to have two people in the same time zone look at these
issues!  :-)

Keith

**Comment by keitharnoldsmith** [comment]
On the subject of possibly related failures, there were a couple mails from
evergreen this morning reporting that test/checkpoint failed.  See, for
example,
https://evergreen.mongodb.com/task/wiredtiger_ubuntu1804_stress_tests_checkpoint_stress_test_b3715b8a005317e6cd98960e9d2d5d0eaa8328bd_20_05_08_05_13_25/0
.

Not sure if it's related or not, but I don't recall seeing this particular
failure until now, and it's also looks like a checkpoint that doesn't
include the correct data. (Although it's a data mismatch not a checksum
mismatch.)

Keith

**Comment by sueloverso** [comment]
@keithbostic turns out that these changes (as in the whole thing, not your recent additions) break incremental backup. We go from copying the entire file (correct but extra work) to not copying enough, apparently, and getting corruption with 0 blocks. I'm still digging.

**Comment by keithbostic** [comment]
@sueloverso, I'm not terribly shocked it broke. I stared at the code, and I thought we were copying everything we needed, but obviously not. If you want to point me at your test case, I'm happy to dig in.

**Comment by sueloverso** [comment]
@keithbostic the repro case is `test_incr_backup -S 2909313755115936578`. It fails quickly and every time on table `t18-0`. You can get more information from the test by adding a `-v 6` to the command line also. I am going to re-instate some of my original ticket debugging. Let me know if you're interested in that.

File: src/block/block_ckpt.c [hunk 1]
```diff
@@ -803,29 +842,34 @@ __ckpt_update(
...
     /*
-     * If this is the live system, we need to record the list of blocks written for this checkpoint
-     * (including the blocks we allocated to write the extent lists).
+     * Record the blocks allocated to write the extent lists. We must record blocks in the live
+     * system's extent lists, as those blocks are a necessary part of the checkpoint a hot backup
+     * might recover. Update blocks in extent lists used to rewrite other checkpoints (for example,
+     * an intermediate checkpoint rewritten because a checkpoint was rolled into it), even though
+     * it's not necessary: those blocks aren't the last checkpoint in the file and so aren't
+     * included in a recoverable checkpoint, they don't matter on a hot backup target until they're
+     * allocated and * used in the context of a live system. Regardless, they shouldn't materially
```
Comment (keitharnoldsmith) [inline src/block/block_ckpt.c:851]: Delete asterisk: "and * used"

**Pull request merged**